http/src/v0/refs.rs

```diff
@@ -138,32 +138,246 @@ async fn refs_paths<T: IpfsTypes>(
+        .try_collect()
+        .await?;
 
     Ok(iplds_refs(ipfs, iplds, max_depth, unique))
 }
 
+#[derive(Debug)]
+pub struct WalkError {
+    pub(crate) last_cid: Cid,
+    pub(crate) reason: WalkFailed,
+}
+
+#[derive(Debug)]
+pub enum WalkFailed {
+    Loading(Error),
+    Parsing(libipld::error::BlockError),
+    DagPb(ipfs::unixfs::ll::ResolveError),
+    IpldWalking(path::WalkFailed),
+}
+
+use std::fmt;
+
+impl fmt::Display for WalkError {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use WalkFailed::*;
+        match &self.reason {
+            Loading(e) => write!(fmt, "loading of {} failed: {}", self.last_cid, e),
+            Parsing(e) => write!(fmt, "failed to parse {} as IPLD: {}", self.last_cid, e),
+            DagPb(e) => write!(
+                fmt,
+                "failed to resolve {} over dag-pb: {}",
+                self.last_cid, e
+            ),
+            // this is asserted in the conformance tests and I don't really want to change the
+            // tests for this
+            IpldWalking(e) => write!(fmt, "{} under {}", e, self.last_cid),
+        }
+    }
+}
+
+impl std::error::Error for WalkError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use WalkFailed::*;
+        match &self.reason {
+            Loading(_) => None, // TODO: anyhow
+            Parsing(e) => Some(e),
+            DagPb(e) => Some(e),
+            IpldWalking(e) => Some(e),
+        }
+    }
+}
+
+impl From<Error> for WalkFailed {
+    fn from(e: Error) -> Self {
+        WalkFailed::Loading(e)
+    }
+}
+
+impl From<libipld::error::BlockError> for WalkFailed {
+    fn from(e: libipld::error::BlockError) -> Self {
+        WalkFailed::Parsing(e)
+    }
+}
+
+impl From<ipfs::unixfs::ll::ResolveError> for WalkFailed {
+    fn from(e: ipfs::unixfs::ll::ResolveError) -> Self {
+        WalkFailed::DagPb(e)
+    }
+}
+
+impl From<path::WalkFailed> for WalkFailed {
+    fn from(e: path::WalkFailed) -> Self {
+        WalkFailed::IpldWalking(e)
+    }
+}
+
+impl From<(WalkFailed, Cid)> for WalkError {
+    fn from((reason, last_cid): (WalkFailed, Cid)) -> Self {
+        WalkError { last_cid, reason }
+    }
+}
+
+/// The IpfsPath walk can end in with the target block loaded or parsed and optionally projected as
+/// an Ipld.
+#[derive(Debug)]
+pub enum Loaded {
+    /// The raw block from `ipfs.get_block`
+    Raw(Box<[u8]>),
+    /// Possibly projected IPLD value.
+    Ipld(Ipld),
+}
+
 /// Walks the `path` while loading the links.
 ///
-/// # Panics
-///
-/// If there are dag-pb nodes and the libipld has changed it's dag-pb tree structure.
+/// Returns the Cid where we ended up, and an optional Ipld structure if one was projected, and the
+/// path inside the last document we walked.
 pub async fn walk_path<T: IpfsTypes>(
     ipfs: &Ipfs<T>,
     mut path: IpfsPath,
-) -> Result<(Cid, Ipld), Error> {
+) -> Result<(Cid, Loaded, Vec<String>), WalkError> {
```

Comment:
This function should become `ipfs::resolve(&self, path: IpfsPath) -> Result<(Cid, Loaded, Vec<String>), _>` in the near-ish future, I am quite confident this is about ready but I would prefer to leave it to another PR as it's a quite large API change at least.